URL option for BaseRunAsSuperuserCommand (#81025)
elastic/elasticsearch: docs/reference/commands/create-enrollment-token.asciidoc

`-E <KeyValuePair>`:: Configures a standard {es} or {xpack} setting.

`-f, --force`:: Forces the command to run against an unhealthy cluster.

`-h, --help`:: Returns all of the command parameters.

`-s, --scope`:: Specifies the scope of the generated token. Supported values are `node` and `kibana`.

`--url`:: Specifies the base URL (hostname and port) that the tool uses to submit API
requests to {es}. The default value is determined from the settings in your
`elasticsearch.yml` file. If `xpack.security.http.ssl.enabled`  is set to `true`,

> you must specify an HTTPS URL.

Reviewer: Do you think it would be preferable that we drop the "If `xpack.security.http.ssl.enabled`  is set to `true`,
you must specify an HTTPS URL" scheme condition, by having the tool check for that itself (like it does for the "default URL")?

Moreover, a URL is more than the scheme and the host and port pair, and using DNS names to refer to the node might not be ideal, as it might not be included in the SAN of the cert.
Do you think it would be preferable that this option be more focused on the **ip and port that the local node can be reached at**?

Author: > Do you think it would be preferable that this option be more focused on the ip and port that the local node can be reached at?

I contemplated the same but in the end settled to a single option as a simpler thing, happy to discuss. 

 >  a URL is more than the scheme and the host and port pair

true but what we should be aiming for here is not naming strictness for the sake of it but a name for the option that’s understandable by (or at least explainable for ) the majority of users. I think that `url` fits the bill. 

> and using DNS names to refer to the node might not be ideal, as it might not be included in the SAN of the cert.

the idea is that if you ever need to use this parameter you do so because you _know_ that you want to talk to the node at an ip or hostname that is in the SANs but CommandLineHttpClient picks another. So you *know* what should be in the url (either a hostname or an IP address ) 

> Do you think it would be preferable that we drop the "If xpack.security.http.ssl.enabled is set to true,
you must specify an HTTPS URL" scheme condition, by having the tool check for that itself (like it does for the "default URL")?

If we change this to two parameters then yes it makes sense , we shouldn’t be asking the users for a third parameter (scheme) but I’m not 100% that ee should change. 

Do we introduce an ip and a port parameter? Do we mandate both if one is passed or have default values for the port ?

Reviewer: > Do we introduce an ip and a port parameter? Do we mandate both if one is passed or have default values for the port ?

I don't have a preference. Having to specify both the ip and the port simultaneously (as is the case with this `url` parameter), as a single option is OK.

----
bin/elasticsearch-create-enrollment-token -s node
----

> The following command creates an enrollment token for enrolling a kibana instance into a cluster, 
> while overriding the url where the local elasticsearch node is reachable:

Reviewer 2: ```suggestion
The following command creates an enrollment token for enrolling a {kib} instance into a cluster, 
and overrides the URL where the local {es} node is reachable:
```

Author: `and` feels off here ( and I mean "feels", I can't actually argue about it being wrong :). To me your suggestion reads as if the example command invocation does two things: 

1. It creates an enrollment token
2. It overrides the URL where the local elasticsearch node is reachable

which is not accurate. 
It does only 1. but 2 is something that temporarily happens - and only for the context of this command's execution - so that 1. can succeed. 

WDYT @lockewritesdocs  ?

Reviewer 2: Ooooo, I see! In that case, let's explore options. Choose whichever one you like better, or let me know neither hits the mark.

This rendition is similar to your original text, but adds "temporarily" to distinguish the time constraint:

>The following command creates an enrollment token for enrolling a {kib} instance into a cluster, while temporarily overriding the URL where the local {es} node is reachable:

Another option dismisses the notion of the interaction between 1 and 2 entirely, and instead focuses on what the URL denotes:

>The following command creates an enrollment token for enrolling a {kib} instance into a cluster. The specified URL indicates where the local {es} node is reachable:

Author: Thanks for the discussion Adam. It actually makes me think that the original "is reachable" is misleading. Setting this option doesn't change at all where elasticsearch is reachable or listens to, but only where the CLI tool will talk to.
In my original version, there was a "while overriding the url where _the CLI tool thinks that_ the local elasticsearch node is reachable" implied there. 

With that context, do you have any suggestions on how to frame this ?

Reviewer 2: Given that context, I think something like the following statement works:

>The following command creates an enrollment token for enrolling a {kib} instance into a cluster. The specified URL indicates where the `elasticsearch-create-enrollment-token` tool attempts to reach the local {es} node:

If we want to indicate more about "where the CLI tool thinks that the local ES node is reachable," then we might need further clarification of what happens if the tool doesn't find the node at the specified URL:

>The following command creates an enrollment token for enrolling a {kib} instance into a cluster. The specified URL indicates where the `elasticsearch-create-enrollment-token` tool attempts to reach the local {es} node. If the tool can't reach {es} at the specified port, it iterates through port numbers sequentially on the indicated host:

Caveat: I'm not certain that the last sentence here ☝️ is true.

Author: > Caveat: I'm not certain that the last sentence here point_up is true.

It doesn't work like this actually, 

> The following command creates an enrollment token for enrolling a {kib} instance into a cluster. The specified URL indicates where the elasticsearch-create-enrollment-token tool attempts to reach the local {es} node:

sounds good to me, let's keep this simple
